### PR TITLE
Improved solution

### PR DESCRIPTION
**Pull request by randyheydon**
This pull request addresses all of the tests that were added in pull request #3 (this pull request also includes those tests).  This makes the DfM checks closer match manufacturing capabilities.

Note that one of the new checks, for small cuts, causes the code to run much slower than it did before.  All of the `distToShape` calls take roughly two-thirds of the full run time of the test suite.  I plan to spend some time tomorrow trying to cut down on the number of those calls.  But if all else fails, the code now at least works correctly.

### CODE REVIEW

**Comment by randyheydon** [comment]
I managed to cut down all the `distToShape` calls, so the solution time is now close to where it was before.  I'm now pretty happy with the results, so please review at your leisure.

**Comment by 10flow** [comment]
Sound awesome! I'm going to walk through the code and try it out. I'll get back with any comments and questions in the next couple days.

File: paperlesscad/solution.py [hunk 1]
```diff
@@ -49,111 +55,247 @@ def dfm_check(step_path):
...
+            if is_tight_corner:
+                continue
+            # Precise distance checks are expensive, so first check if the
+            # bounding boxes of the two surfaces are close enough to even
+            # potentially be a concern.
+            bound_box1 = f1.BoundBox
+            bound_box1.enlarge(0.5 * kerf_width)
+            bound_box2 = f2.BoundBox
+            bound_box2.enlarge(0.5 * kerf_width)
+            if not bound_box1.intersect(bound_box2):
```
Comment (10flow) [inline paperlesscad/solution.py]: Profiled this out of curiosity... `intersect()` is an insignificant part of total runtime. If you had too many `intersect` calls, I would suggest looking at `Shape.proximity()`.

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.003    0.003    1.563    1.563 cut.py:1(<module>)
      100    0.177    0.002    1.471    0.015 solution.py:26(dfm_check)
      100    0.969    0.010    0.969    0.010 {method 'read' of 'Part.Shape' objects}
      100    0.098    0.001    0.098    0.001 {method 'distToShape' of 'Part.Shape' objects}
        1    0.000    0.000    0.084    0.084 solution.py:8(<module>)
        1    0.075    0.075    0.084    0.084 utils.py:1(<module>)
     1800    0.046    0.000    0.046    0.000 {method 'normalAt' of 'Part.Face' objects}
      300    0.034    0.000    0.035    0.000 solution.py:42(<genexpr>)
    23600    0.014    0.000    0.018    0.000 utils.py:20(__init__)
     1100    0.018    0.000    0.018    0.000 solution.py:40(<genexpr>)
      100    0.015    0.000    0.016    0.000 solution.py:59(<genexpr>)
      100    0.015    0.000    0.016    0.000 solution.py:48(<genexpr>)
      100    0.015    0.000    0.016    0.000 solution.py:53(<genexpr>)
    20200    0.011    0.000    0.014    0.000 utils.py:24(__eq__)
      100    0.012    0.000    0.013    0.000 solution.py:46(<genexpr>)
     3000    0.012    0.000    0.012    0.000 {method 'parameter' of 'Part.GeomSurface' objects}
        1    0.000    0.000    0.009    0.009 <string>:34(<module>)
        1    0.002    0.002    0.009    0.009 <string>:37(InitApplications)
       15    0.000    0.000    0.005    0.000 re.py:192(compile)
       15    0.000    0.000    0.005    0.000 re.py:230(_compile)
       15    0.000    0.000    0.005    0.000 sre_compile.py:567(compile)
      100    0.002    0.000    0.005    0.000 solution.py:68(<dictcomp>)
        1    0.000    0.000    0.005    0.005 platform.py:10(<module>)
     4800    0.002    0.000    0.005    0.000 {method 'update' of 'set' objects}
     7400    0.004    0.000    0.005    0.000 utils.py:9(is_close)
        1    0.000    0.000    0.004    0.004 __init__.py:45(<module>)
    23600    0.004    0.000    0.004    0.000 {method 'hashCode' of 'Part.Shape' objects}
    20200    0.003    0.000    0.003    0.000 {method 'isSame' of 'Part.Shape' objects}
     7139    0.003    0.000    0.003    0.000 {isinstance}
     2000    0.003    0.000    0.003    0.000 {method 'intersect' of 'Base.BoundBox' objects}
```

File: paperlesscad/solution.py [hunk 2]
```diff
@@ -49,111 +55,247 @@ def dfm_check(step_path):
...
-    leftovers = set(faces) - planes - cylinders - cones
+    bad_surfaces = set(
+        f for f in faces
+        if isinstance(f.Surface, (Part.Sphere, Part.Toroid)))
+    leftovers = set(faces) - planes - cylinders - cones - bad_surfaces
+
+    # Construct map of which faces are connected to a given edge.
+    # There is usually two faces per edge, but sometimes only one.  I don't
+    # think that more than two is possible, but let's not assume.
+    # NOTE FreeCAD 0.17 will have an "ancestorsOfType" method that could
```
Comment (10flow) [inline paperlesscad/solution.py]: We have a hotrod version of FreeCAD 0.17 + Python3 running, and `ancestorsOfType` does in fact work.

File: paperlesscad/solution.py [hunk 3]
```diff
@@ -49,111 +55,247 @@ def dfm_check(step_path):
...
+        for e in f.Edges:
+            edge_to_faces[HashShape(e)].add(HashShape(f))
+
+    # Construct map of which faces are adjacent to a given face (i.e. all faces
+    # that share an edge with this face).
+    face_to_faces = {HashShape(f): set() for f in faces}
+    for f in faces:
+        for e in f.Edges:
+            face_to_faces[HashShape(f)].update(edge_to_faces[HashShape(e)])
+    for k, v in face_to_faces.iteritems():
```
Comment (10flow) [inline paperlesscad/solution.py]: `items()` for Python3 compatibility?

File: paperlesscad/solution.py [hunk 4]
```diff
@@ -49,111 +55,247 @@ def dfm_check(step_path):
...
+    # the attached faces to see if they meet at a significant angle, and
+    # whether they form a convex or concave corner.
+    # NOTE This will not work for an edge that's connected to only one face.
+    # If that face connects to itself with a sharp angle, this code will not
+    # detect that properly.  An edge connected to more than two faces is also
+    # ignored because I don't know what that would look like.
+    # The difference between "tight-corner" and "tight-corner-mild" is defined
+    # here as being ten degrees.
+    shallow_limit = 10 * pi/180
+    for e, fs in edge_to_faces.iteritems():
```
Comment (10flow) [inline paperlesscad/solution.py]: items() for Python3 compatibility?

File: paperlesscad/solution.py (hunk 2 above)
Comment (randyheydon) [inline paperlesscad/solution.py]: Cool, I didn't realize that Python 3 support in FreeCAD is ready for testing FreeCAD.  Is the plan to use FreeCAD 0.17 in production in the near future?  If so, there are a few small cleanups that could be made (both `ancestorsOfType` and `HashableShape` are available in it).

File: paperlesscad/solution.py (hunk 3 above)
Comment (randyheydon) [inline paperlesscad/solution.py]: Certainly could be.  There are few places that use iter* methods that could be changed without any effect on behaviour.

File: paperlesscad/solution.py (hunk 1 above)
Comment (randyheydon) [inline paperlesscad/solution.py]: Oh yeah, `intersect` is fast, which is why it's here :).  You'll notice that `distToShape` is still pretty high on the list, despite its low number of calls, and it used to be higher before `intersect` replaced most of it.

**Pull request merged**